### PR TITLE
feat: new action: lock screen

### PR DESCRIPTION
**Pull request by yzqzss**
On Android 9+, we can lock the screen through accessibility service.

image

close: #51

### CODE REVIEW

**Comment by jrpie** [comment]
Nice, thank you!

**Comment by jrpie** [comment]
But woudn't it be cleaner to implement this through the [device admin API](https://developer.android.com/work/device-admin)?

The [guide on accessibility services](https://developer.android.com/guide/topics/ui/accessibility/service) explicitely says:

> Note: Your app must use platform-level accessibility services only for the purpose of helping users with disabilities interact with your app.

**Comment by jrpie** [comment]
[DevicePolicyManager#lockNow()](https://developer.android.com/reference/android/app/admin/DevicePolicyManager#lockNow()) would furthermore also work on older devices.

However the description doesn't sound ideal either:
> This method secures the device in response to an urgent situation, such as a lost or stolen device. After this method is called, the device must be unlocked using strong authentication (PIN, pattern, or password). This API is intended for use only by device admins.

**Comment by yzqzss** [comment]
> The [guide on accessibility services](https://developer.android.com/guide/topics/ui/accessibility/service) explicitely says:
> 
> > Note: Your app must use platform-level accessibility services only for the purpose of helping users with disabilities interact with your app.

Using accessibility to lock screen is indeed an abuse, but sadly, there is no better way to lock screen without root. :(

**Comment by yzqzss** [comment]
> However the description doesn't sound ideal either:
> 
> > This method secures the device in response to an urgent situation, such as a lost or stolen device. After this method is called, the device must be unlocked using strong authentication (PIN, pattern, or password). This API is intended for use only by device admins.

Yeah, so I didn't consider using the device (Policy) manager to achieve this function.

**Comment by jrpie** [comment]
But the DevicePolicyManager would also work without root, right?
So as far as I understand there is a choice between two hacks. (I'm not opposed to hacks, and there is much worse in the codebase, I'd just want to avoid hacks when possible).

However afaik google is quite picky about what uses of accessibility services they allow in the playstore (see e.g. [this](https://support.google.com/googleplay/android-developer/answer/10964491?hl=en))

(Perhaps I should just give up on play store, f-droid is much better and I don't like doing unpaid work for google - especially all the bureaucracy they require - on the other hand play store is quite large and I think open source apps being available there is important for users to even discover that alternatives exist)

**Comment by yzqzss** [comment]
The top 4 launchers on Play Store all have the lock screen feature.

- [Nova Launcher](https://play.google.com/store/apps/details?id=com.teslacoilsw.launcher) Accessibility, Device Admin, Root
- [Microsoft Launcher](https://play.google.com/store/apps/details?id=com.microsoft.launcher) Accessibility, Device Admin
- [Smart Launcher](https://play.google.com/store/apps/details?id=ginlemon.flowerfree) Accessibility
- [Niagara Launcher](https://play.google.com/store/apps/details?id=bitpit.launcher) Accessibility

Maybe Google isn't that picky about the uses? IDK, I'm an F-Droid user. :)

**Comment by jrpie** [comment]
I didn't know that this hack is so common, thanks for doing the research!

File: app/src/main/res/xml/accessibility_service_config.xml
Comment (jrpie) [inline app/src/main/res/xml/accessibility_service_config.xml:1]: Is there any reason to use these very generic masks? Especially `android:accessibilityEventTypes="typeAllMask"` looks unnecessary to me. (Unfortunately there doesn't seem to be a "nothing" mask)

**Comment by jrpie** [comment]
image
This sounds like a privacy nightmare to me.

**Comment by jrpie** [comment]
That looks more reasonable to me:
image

I am very sorry that so much work has been in vain, but ultimately I think the device admin approach is better as it
(a) works on older devices as well and
(b) has a better interface to specify what functionality we actually need

**Pull request merged**